File: GamePhysicsSdk/EasyPhisicsDotNet/EpxConvexConvexContact.cs
using System;

namespace EasyPhisicsDotNet
{
    class EpxConvexConvexContact
    {
        /// <summary>
        /// 凸メッシュの衝突パターン?
        /// </summary>
        private enum EpxSatType
        {
            /// <summary>凸メッシュAの頂点と、凸メッシュBの面が衝突している</summary>
            PointAFacetB,

            /// <summary>凸メッシュBの頂点と、凸メッシュAの面が衝突している</summary>
            PointBFacetA,

            /// <summary>エッジ同士が衝突している</summary>
            EdgeEdge,
        }

        /// <summary>
        /// ２つの凸メッシュの衝突検出
        /// 
        /// 二つの凸メッシュの並進速度、角速度は使用していないことに注意。
        /// </summary>
        /// <param name="convexA">凸メッシュA</param>
        /// <param name="transformA">Aのワールド変換行列</param>
        /// <param name="convexB">凸メッシュB</param>
        /// <param name="transformB">Bのワールド変換行列</param>
        /// <param name="normal">[out] 衝突点の法線ベクトル（ワールド座標系）</param>
        /// <param name="penetrationDepth">[out]貫通深度</param>
        /// <param name="contactPointA">[out]衝突点（剛体Aのローカル座標系）</param>
        /// <param name="contactPointB">[out]衝突点（剛体Bのローカル座標系）</param>
        /// <returns>衝突検出ができたかどうか(衝突しているかどうか)</returns>
        public static bool Execute(
            EpxConvexMesh convexA,
            EpxTransform3 transformA,
            EpxConvexMesh convexB,
            EpxTransform3 transformB,
            out EpxVector3 normal,
            out float penetrationDepth,
            out EpxVector3 contactPointA,
            out EpxVector3 contactPointB)
        {
            // 座標系変換の回数を減らすため、面数の多い方を座標系の基準にとる
            if (convexA.m_facets.Count >= convexB.m_facets.Count)
            {
                return _CheckContact(
                    convexA, transformA,
                    convexB, transformB,
                    out normal, out penetrationDepth,
                    out contactPointA, out contactPointB);
            }
            else
            {
                var ret = _CheckContact(
                    convexB, transformB,
                    convexA, transformA,
                    out normal, out penetrationDepth,
                    out contactPointB, out contactPointA);
                normal = -normal;
                return ret;
            }
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="convexA">凸メッシュA</param>
        /// <param name="transformA">Aのワールド変換行列</param>
        /// <param name="convexB">凸メッシュB</param>
        /// <param name="transformB">Bのワールド変換行列</param>
        /// <param name="normal">[out] 衝突点の法線ベクトル（ワールド座標系）</param>
        /// <param name="penetrationDepth">[out]貫通深度</param>
        /// <param name="contactPointA">[out]衝突点（剛体Aのローカル座標系）</param>
        /// <param name="contactPointB">[out]衝突点（剛体Bのローカル座標系）</param>
        /// <returns>衝突検出ができたかどうか(衝突しているかどうか)</returns>
        private static bool _CheckContact(
            EpxConvexMesh convexA,
            EpxTransform3 transformA,
            EpxConvexMesh convexB,
            EpxTransform3 transformB,
            out EpxVector3 normal,
            out float penetrationDepth,
            out EpxVector3 contactPointA,
            out EpxVector3 contactPointB)
        {
            // 最も浅い貫通深度とそのときの分離軸
            float distanceMin;
            EpxVector3 separationAxis;
            EpxSatType satType;
            bool axisFlip;

            // 分離軸判定
            if (!_DetectSeparatingAxis(convexA, transformA, convexB, transformB, out distanceMin, out separationAxis, out satType, out axisFlip))
            {
                normal = null;
                penetrationDepth = 0;
                contactPointA = null;
                contactPointB = null;
                return false;
            }

            // ここまで到達したので、２つの凸メッシュは交差している。
            // また、反発ベクトル(axisMin)と貫通深度(distanceMin)が求まった。
            // 反発ベクトルはＡを押しだす方向をプラスにとる。

            // Bローカル→Aローカルへの変換
            var tAB = transformA.orthoInverse() * transformB;

            // Aローカル→Bローカルへの変換
            var tBA = tAB.orthoInverse();

            //----------------------------------------------------------------------------
            // 衝突座標検出

            int collCount = 0;

            var closestMinSqr = float.MaxValue;
            var closestPointA = new EpxVector3(0, 0, 0); // 初期値はこれでよいのかよくわからない。(秋山)
            var closestPointB = new EpxVector3(0, 0, 0); // 初期値はこれでよいのかよくわからない。(秋山)
            var separation = 1.1f * Math.Abs(distanceMin) * separationAxis;

            foreach (var facetA in convexA.m_facets)
            {
                var checkA = facetA.Normal.dot(-separationAxis);
                if (satType == EpxSatType.PointBFacetA && checkA < 0.99f && axisFlip)
                    continue;                    // 判定軸が面Aの法線のとき、向きの違うAの面は判定しない

                if (checkA < 0.0f)
                    continue;                    // 衝突面と逆に向いている面は判定しない

                foreach (var facetB in convexB.m_facets)
                {
                    float checkB = facetB.Normal.dot(tBA.Rotate(separationAxis));
                    if (satType == EpxSatType.PointAFacetB && checkB < 0.99f && !axisFlip)
                        continue;                        // 判定軸が面Bの法線のとき、向きの違うBの面は判定しない

                    if (checkB < 0.0f)
                        continue;                        // 衝突面と逆に向いている面は判定しない

                    collCount++;

                    // 面Ａと面Ｂの最近接点を求める
                    var triangleA = new EpxVector3[] {
                        separation + convexA.Vertices[facetA.vertId[0]],
                        separation + convexA.Vertices[facetA.vertId[1]],
                        separation + convexA.Vertices[facetA.vertId[2]],
                    };

                    var triangleB = new EpxVector3[] {
                        tAB.Transform(convexB.Vertices[facetB.vertId[0]]),
                        tAB.Transform(convexB.Vertices[facetB.vertId[1]]),
                        tAB.Transform(convexB.Vertices[facetB.vertId[2]]),
                    };

                    // エッジ同士の最近接点算出
                    for (int i = 0; i < 3; i++)
                    {
                        for (int j = 0; j < 3; j++)
                        {
                            EpxVector3 sA;
                            EpxVector3 sB;
                            EpxClosestFunction.epxGetClosestTwoSegments(
                                triangleA[i], triangleA[(i + 1) % 3],
                                triangleB[j], triangleB[(j + 1) % 3],
                                out sA, out sB);

                            var dSqr = (sA - sB).LengthSqr();
                            if (dSqr < closestMinSqr)
                            {
                                closestMinSqr = dSqr;
                                closestPointA = sA;
                                closestPointB = sB;
                            }
                        }
                    }

                    // 頂点Ａ→面Ｂの最近接点算出
                    for (int i = 0; i < 3; i++)
                    {
                        EpxVector3 s;
                        EpxClosestFunction.epxGetClosestPointTriangle(triangleA[i], triangleB[0], triangleB[1], triangleB[2], tBA.Rotate(facetB.Normal), out s);
                        float dSqr = (triangleA[i] - s).LengthSqr();
                        if (dSqr < closestMinSqr)
                        {
                            closestMinSqr = dSqr;
                            closestPointA = triangleA[i];
                            closestPointB = s;
                        }
                    }

                    // 頂点Ｂ→面Ａの最近接点算出
                    for (int i = 0; i < 3; i++)
                    {
                        EpxVector3 s;
                        EpxClosestFunction.epxGetClosestPointTriangle(triangleB[i], triangleA[0], triangleA[1], triangleA[2], facetA.Normal, out s);
                        float dSqr = (triangleB[i] - s).LengthSqr();
                        if (dSqr < closestMinSqr)
                        {
                            closestMinSqr = dSqr;
                            closestPointA = s;
                            closestPointB = triangleB[i];
                        }
                    }
                }
            }

            normal = transformA.Rotate(separationAxis);
            penetrationDepth = distanceMin;
            contactPointA = closestPointA - separation;
            contactPointB = tBA.Transform(closestPointB);

            return true;
        }

        private static bool _DetectSeparatingAxis(
            EpxConvexMesh convexA,
            EpxTransform3 transformA,
            EpxConvexMesh convexB,
            EpxTransform3 transformB,
            out float distanceMin, out EpxVector3 axisMin, out EpxSatType satType, out bool axisFlip)
        {
            distanceMin = -float.MaxValue;
            axisMin = new EpxVector3(0.0f);
            satType = EpxSatType.EdgeEdge;
            axisFlip = false;  // 初期値がfalseでよいのかわからない。元のC++でも初期化されない?(秋山)

            // Bローカル→Aローカルへの変換
            var tAB = transformA.orthoInverse() * transformB;

            // Aローカル→Bローカルへの変換
            var tBA = tAB.orthoInverse();

            #region Aの各面へのBの頂点のめり込みをチェック
            foreach (var facet in convexA.m_facets)
            {
                // Aの面法線を分離軸とする
                var separatingAxis = facet.Normal;
                // A面を分離軸に投影
                float maxA = separatingAxis.dot(convexA.Vertices[facet.vertId[0]]);
                // ConvexBを分離軸に投影(必要なのは最小値のみ)
                float minB;
                EpxConvexMeshUtil.GetProjection(convexB, tBA.Rotate(facet.Normal), out minB);
                var offset = tAB.Translation.dot(separatingAxis);
                minB += offset;

                // 判定
                var d = minB - maxA;
                if (d > 0)
                    return false;
                if (d > distanceMin)
                {
                    distanceMin = d;
                    axisMin = -separatingAxis;
                    axisFlip = true;
                    satType = EpxSatType.PointBFacetA;
                }
            }
            #endregion Aの各面へのBの頂点のめり込みをチェック

            #region  Bの各面へのAの頂点のめり込みをチェック
            foreach (EpxFacet facet in convexB.m_facets)
            {
                // ConvexBの面法線をConvecAの座標系に変換したものを分離軸とする
                var separatingAxis = tAB.Rotate(facet.Normal);
                // B面を分離軸に投影
                float maxB = separatingAxis.dot(convexB.Vertices[facet.vertId[0]]);
                float offset = tAB.Translation.dot(separatingAxis);
                maxB += offset;

                // ConvexAを分離軸に投影
                float minA;
                EpxConvexMeshUtil.GetProjection(convexA, separatingAxis, out minA);

                // 判定
                var d = minA - maxB;
                if (d > 0)
                    return false;
                if (d > distanceMin)
                {
                    distanceMin = d;
                    axisMin = separatingAxis;
                    axisFlip = false;
                    satType = EpxSatType.PointAFacetB;
                }
            }
            #endregion Bの各面へのAの頂点のめり込みをチェック

            #region ConvexAとConvexBのエッジの外積を分離軸とする
            foreach (var edgeA in convexA.Edges)
            {
                var edgeVecA = convexA.Vertices[edgeA.V1] - convexA.Vertices[edgeA.V0];
                foreach (var edgeB in convexB.Edges)
                {
                    var edgeVecB = tAB.Rotate(convexB.Vertices[edgeB.V1] - convexB.Vertices[edgeB.V0]);

                    EpxVector3 separatingAxis = edgeVecA.cross(edgeVecB);
                    if (separatingAxis.LengthSqr() < float.Epsilon * float.Epsilon)
                        continue;

                    separatingAxis = separatingAxis.Normalize();

                    // ConvexAを分離軸に投影
                    float minA, maxA;
                    EpxConvexMeshUtil.GetProjection(convexA, separatingAxis, out minA, out maxA);

                    // ConvexBを分離軸に投影
                    float minB, maxB;
                    EpxConvexMeshUtil.GetProjection(convexB, tBA.Rotate(separatingAxis), out minB, out maxB);
                    float offset = tAB.Translation.dot(separatingAxis);
                    minB += offset;
                    maxB += offset;

                    // 判定
                    if (!EPX_CHECK_MINMAX(separatingAxis, minA, maxA, minB, maxB,
                        ref distanceMin, ref axisMin, ref axisFlip))
                        return false;

                    satType = EpxSatType.EdgeEdge;
                }
            }
            #endregion ConvexAとConvexBのエッジの外積を分離軸とする

            return true;
        }


        /// <summary>
        /// </summary>
        /// <param name="axis">分離軸</param>
        /// <param name="minA">分離軸に投影された凸メッシュAの最小値</param>
        /// <param name="maxA">分離軸に投影された凸メッシュAの最大値</param>
        /// <param name="minB">分離軸に投影された凸メッシュBの最小値</param>
        /// <param name="maxB">分離軸に投影された凸メッシュBの最大値</param>
        /// <param name="satCount">[io]衝突の数?</param>
        /// <param name="distanceMin">Aの面へのめり込みの深さ</param>
        /// <param name="axisMin"></param>
        /// <param name="axisFlip"></param>
        /// <returns>凸メッシュAとBが衝突しているかどうか</returns>
        static bool EPX_CHECK_MINMAX(EpxVector3 axis, float minA, float maxA, float minB, float maxB,
            ref float distanceMin, ref EpxVector3 axisMin, ref bool axisFlip)
        {
            var d1 = minA - maxB;
            if (d1 > 0)
                return false;   // BはAの面の裏側に離れて存在存在していて重複はしていない

            var d2 = minB - maxA;   // Bの頂点がAの面にめり込んでいる深さ
            if (d2 > 0)
                return false;   // BはAの面の表側に離れて存在していて重複はしていない

            // この判定必要???
            if (d1 > distanceMin)
            {
                distanceMin = d1;
                axisMin = axis;
                axisFlip = false;
            }

            // Aの面へのめり込みがもっとも浅いものを選択する
            if (d2 > distanceMin)
            {
                distanceMin = d2;
                axisMin = -axis;
                axisFlip = true;
            }

            return true;
        }

    }
}

File: GamePhysicsSdk/easy_physics/collision/EpxConvexConvexContact.cpp
/*
	Copyright (c) 2012 Hiroshi Matsuike

	This software is provided 'as-is', without any express or implied
	warranty. In no event will the authors be held liable for any damages
	arising from the use of this software.

	Permission is granted to anyone to use this software for any purpose,
	including commercial applications, and to alter it and redistribute it
	freely, subject to the following restrictions:

	1. The origin of this software must not be misrepresented; you must not
	claim that you wrote the original software. If you use this software
	in a product, an acknowledgment in the product documentation would be
	appreciated but is not required.

	2. Altered source versions must be plainly marked as such, and must not be
	misrepresented as being the original software.

	3. This notice may not be removed or altered from any source distribution.
*/

#include "EpxConvexConvexContact.h"
#include "EpxClosestFunction.h"

namespace EasyPhysics {

#ifndef EPX_DOXYGEN_SKIP

enum EpxSatType {
	EpxSatTypePointAFacetB,
	EpxSatTypePointBFacetA,
	EpxSatTypeEdgeEdge,
};

#define EPX_CHECK_MINMAX(axis,AMin,AMax,BMin,BMax,type) \
{\
	satCount++;\
	EpxFloat d1 = AMin - BMax;\
	EpxFloat d2 = BMin - AMax;\
	if(d1 >= 0.0f || d2 >= 0.0f) {\
		return false;\
	}\
	if(distanceMin < d1) {\
		distanceMin = d1;\
		axisMin = axis;\
		satType = type;\
		axisFlip = false;\
	}\
	if(distanceMin < d2) {\
		distanceMin = d2;\
		axisMin = -axis;\
		satType = type;\
		axisFlip = true;\
	}\
}

EpxBool epxConvexConvexContact_local(
	const EpxConvexMesh &convexA,const EpxTransform3 &transformA,
	const EpxConvexMesh &convexB,const EpxTransform3 &transformB,
	EpxVector3 &normal,
	EpxFloat &penetrationDepth,
	EpxVector3 &contactPointA,
	EpxVector3 &contactPointB)
{
	EpxTransform3 transformAB,transformBA;
	EpxMatrix3 matrixAB,matrixBA;
	EpxVector3 offsetAB,offsetBA;
	
	// Bローカル→Aローカルへの変換
	transformAB = orthoInverse(transformA) * transformB;
	matrixAB = transformAB.getUpper3x3();
	offsetAB = transformAB.getTranslation();
	
	// Aローカル→Bローカルへの変換
	transformBA = orthoInverse(transformAB);
	matrixBA = transformBA.getUpper3x3();
	offsetBA = transformBA.getTranslation();
	
	// 最も浅い貫通深度とそのときの分離軸
	EpxFloat distanceMin = -EPX_FLT_MAX;
	EpxVector3 axisMin(0.0f);
	EpxSatType satType = EpxSatTypeEdgeEdge;
	EpxBool axisFlip;
	
	//----------------------------------------------------------------------------
	// 分離軸判定
	
	int satCount = 0;

	// ConvexAの面法線を分離軸とする
	for(EpxUInt32 f=0;f<convexA.m_numFacets;f++) {
		const EpxFacet &facet = convexA.m_facets[f];
		const EpxVector3 separatingAxis = facet.normal;

		// ConvexAを分離軸に投影
		EpxFloat minA,maxA;
		epxGetProjection(minA,maxA,&convexA,separatingAxis);
		
		// ConvexBを分離軸に投影
		EpxFloat minB,maxB;
		epxGetProjection(minB,maxB,&convexB,matrixBA * facet.normal);
		EpxFloat offset = dot(offsetAB,separatingAxis);
		minB += offset;
		maxB += offset;
		
		// 判定
		EPX_CHECK_MINMAX(separatingAxis,minA,maxA,minB,maxB,EpxSatTypePointBFacetA);
	}
		
	// ConvexBの面法線を分離軸とする
	for(EpxUInt32 f=0;f<convexB.m_numFacets;f++) {
		const EpxFacet &facet = convexB.m_facets[f];
		const EpxVector3 separatingAxis = matrixAB * facet.normal;
		
		// ConvexAを分離軸に投影
		EpxFloat minA,maxA;
		epxGetProjection(minA,maxA,&convexA,separatingAxis);
		
		// ConvexBを分離軸に投影
		EpxFloat minB,maxB;
		epxGetProjection(minB,maxB,&convexB,facet.normal);
		EpxFloat offset = dot(offsetAB,separatingAxis);
		minB += offset;
		maxB += offset;
		
		// 判定
		EPX_CHECK_MINMAX(separatingAxis,minA,maxA,minB,maxB,EpxSatTypePointAFacetB);
	}
		
	// ConvexAとConvexBのエッジの外積を分離軸とする
	for(EpxUInt32 eA=0;eA<convexA.m_numEdges;eA++) {
		const EpxEdge &edgeA = convexA.m_edges[eA];
		if(edgeA.type != EpxEdgeTypeConvex) continue;

		const EpxVector3 edgeVecA = convexA.m_vertices[edgeA.vertId[1]] - convexA.m_vertices[edgeA.vertId[0]];
		
		for(EpxUInt32 eB=0;eB<convexB.m_numEdges;eB++) {
			const EpxEdge &edgeB = convexB.m_edges[eB];
			if(edgeB.type != EpxEdgeTypeConvex) continue;
				
			const EpxVector3 edgeVecB = matrixAB * (convexB.m_vertices[edgeB.vertId[1]] - convexB.m_vertices[edgeB.vertId[0]]);
			
			EpxVector3 separatingAxis = cross(edgeVecA,edgeVecB);
			if(lengthSqr(separatingAxis) < EPX_EPSILON*EPX_EPSILON) continue;
			
			separatingAxis = normalize(separatingAxis);
			
			// ConvexAを分離軸に投影
			EpxFloat minA,maxA;
			epxGetProjection(minA,maxA,&convexA,separatingAxis);
				
			// ConvexBを分離軸に投影
			EpxFloat minB,maxB;
			epxGetProjection(minB,maxB,&convexB,matrixBA * separatingAxis);
			EpxFloat offset = dot(offsetAB,separatingAxis);
			minB += offset;
			maxB += offset;
			
			// 判定
			EPX_CHECK_MINMAX(separatingAxis,minA,maxA,minB,maxB,EpxSatTypeEdgeEdge);
		}
	}
	
	// ここまで到達したので、２つの凸メッシュは交差している。
	// また、反発ベクトル(axisMin)と貫通深度(distanceMin)が求まった。
	// 反発ベクトルはＡを押しだす方向をプラスにとる。
	
	//epxPrintf("sat check count %d\n",satCount);

	//----------------------------------------------------------------------------
	// 衝突座標検出
	
	int collCount = 0;

	EpxFloat closestMinSqr = EPX_FLT_MAX;
	EpxVector3 closestPointA,closestPointB;
	EpxVector3 separation = 1.1f * fabs(distanceMin) * axisMin;

	for(EpxUInt32 fA=0;fA<convexA.m_numFacets;fA++) {
		const EpxFacet &facetA = convexA.m_facets[fA];

		EpxFloat checkA = dot(facetA.normal,-axisMin);
		if(satType == EpxSatTypePointBFacetA && checkA < 0.99f && axisFlip) {
			// 判定軸が面Aの法線のとき、向きの違うAの面は判定しない
			continue;
		}
			
		if(checkA < 0.0f) {
			// 衝突面と逆に向いている面は判定しない
			continue;
		}
			
		for(EpxUInt32 fB=0;fB<convexB.m_numFacets;fB++) {
			const EpxFacet &facetB = convexB.m_facets[fB];

			EpxFloat checkB = dot(facetB.normal,matrixBA * axisMin);
			if(satType == EpxSatTypePointAFacetB && checkB < 0.99f && !axisFlip) {
				// 判定軸が面Bの法線のとき、向きの違うBの面は判定しない
				continue;
			}
			
			if(checkB < 0.0f) {
				// 衝突面と逆に向いている面は判定しない
				continue;
			}

			collCount++;

			// 面Ａと面Ｂの最近接点を求める
			EpxVector3 triangleA[3] = {
				separation + convexA.m_vertices[facetA.vertId[0]],
				separation + convexA.m_vertices[facetA.vertId[1]],
				separation + convexA.m_vertices[facetA.vertId[2]],
			};
			
			EpxVector3 triangleB[3] = {
				offsetAB + matrixAB * convexB.m_vertices[facetB.vertId[0]],
				offsetAB + matrixAB * convexB.m_vertices[facetB.vertId[1]],
				offsetAB + matrixAB * convexB.m_vertices[facetB.vertId[2]],
			};
			
			// エッジ同士の最近接点算出
			for(int i=0;i<3;i++) {
				if(convexA.m_edges[facetA.edgeId[i]].type != EpxEdgeTypeConvex) continue;
				
				for(int j=0;j<3;j++) {
					if(convexB.m_edges[facetB.edgeId[j]].type != EpxEdgeTypeConvex) continue;

					EpxVector3 sA,sB;
					epxGetClosestTwoSegments(
						triangleA[i],triangleA[(i+1)%3],
						triangleB[j],triangleB[(j+1)%3],
						sA,sB);
						
					EpxFloat dSqr = lengthSqr(sA-sB);
					if(dSqr < closestMinSqr) {
						closestMinSqr = dSqr;
						closestPointA = sA;
						closestPointB = sB;
					}
				}
			}
				
			// 頂点Ａ→面Ｂの最近接点算出
			for(int i=0;i<3;i++) {
				EpxVector3 s;
				epxGetClosestPointTriangle(triangleA[i],triangleB[0],triangleB[1],triangleB[2],matrixAB * facetB.normal,s);
				EpxFloat dSqr = lengthSqr(triangleA[i]-s);
				if(dSqr < closestMinSqr) {
					closestMinSqr = dSqr;
					closestPointA = triangleA[i];
					closestPointB = s;
				}
			}
				
			// 頂点Ｂ→面Ａの最近接点算出
			for(int i=0;i<3;i++) {
				EpxVector3 s;
				epxGetClosestPointTriangle(triangleB[i],triangleA[0],triangleA[1],triangleA[2],facetA.normal,s);
				EpxFloat dSqr = lengthSqr(triangleB[i]-s);
				if(dSqr < closestMinSqr) {
					closestMinSqr = dSqr;
					closestPointA = s;
					closestPointB = triangleB[i];
				}
			}
		}
	}

	//epxPrintf("intersection check count %d\n",collCount);

	normal = transformA.getUpper3x3() * axisMin;
	penetrationDepth = distanceMin;
	contactPointA = closestPointA - separation;
	contactPointB = offsetBA + matrixBA * closestPointB;

	return true;
}

#endif // EPX_DOXYGEN_SKIP

EpxBool epxConvexConvexContact(
	const EpxConvexMesh &convexA,const EpxTransform3 &transformA,
	const EpxConvexMesh &convexB,const EpxTransform3 &transformB,
	EpxVector3 &normal,
	EpxFloat &penetrationDepth,
	EpxVector3 &contactPointA,
	EpxVector3 &contactPointB)
{
	// 座標系変換の回数を減らすため、面数の多い方を座標系の基準にとる

	EpxBool ret;
	if(convexA.m_numFacets >= convexB.m_numFacets) {
		ret = epxConvexConvexContact_local(
			convexA,transformA,
			convexB,transformB,
			normal,penetrationDepth,contactPointA,contactPointB);
	}
	else {
		ret = epxConvexConvexContact_local(
			convexB,transformB,
			convexA,transformA,
			normal,penetrationDepth,contactPointB,contactPointA);
		normal = -normal;
	}

	return ret;
}

} // namespace EasyPhysics
